symengine/polynomial_multivariate.cpp

```diff
@@ -166,35 +162,34 @@ vec_basic MultivariatePolynomial::get_args() const
 
 std::size_t MultivariatePolynomial::__hash__() const
 {
-    std::hash<std::string> hash_string;
     std::size_t seed = MULTIVARIATEPOLYNOMIAL;
     for (auto var : vars_)
-        // boost's method for combining hashes
-        seed ^= hash_string(var->__str__()) + 0x9e3779b + (seed << 6)
-                + (seed >> 2);
+        hash_combine<std::string>(seed, var->__str__());
 
-    std::vector<vec_int> v = order_umap<vec_int, umap_vec_expr>(dict_);
-
-    for (auto vec : v) {
-        seed ^= vec_int_hash()(dict_.find(vec)->first) + 0x9e3779b + (seed << 6)
-                + (seed >> 2);
-        seed ^= (dict_.find(vec)->second).get_basic()->__hash__() + 0x9e3779b
-                + (seed << 6) + (seed >> 2);
+    for (auto &p : dict_) {
+        std::size_t t = vec_int_hash()(p.first);
+        hash_combine<Basic>(t, *(p.second.get_basic()));
+        seed ^= t;
     }
     return seed;
 }
 
 int MultivariatePolynomial::compare(const Basic &o) const
 {
-    // copied from UIntPoly::compare and then modified.
+    // copied from UnivariateIntPolynomial::compare and then modified.
     const MultivariatePolynomial &s
         = static_cast<const MultivariatePolynomial &>(o);
 
-    int cmp = set_compare<set_basic>(vars_, s.vars_);
+    if (vars_.size() != s.vars_.size())
+        return vars_.size() < s.vars_.size() ? -1 : 1;
+    if (dict_.size() != s.dict_.size())
+        return dict_.size() < s.dict_.size() ? -1 : 1;
+
+    int cmp = unified_compare(vars_, s.vars_);
```

Comment:
I think such comparison should be handled in a better way, e.g. 
```
return unified_compare(std::forward_as_tuple(vars_.size(), dict_.size(), vars_, dict_),
                std::forward_as_tuple(s.vars_.size(), s.dict_.size(), s.vars_, s.dict_));
```

A demo has just been implemented: https://github.com/htfy96/cpp-cmp/